File: backend/ingestion/yahoo.py
"""
Yahoo Finance data ingestion via yfinance.
Fetches historical prices and fundamental data for stocks.
"""

import logging
from datetime import date, datetime
from typing import Optional

import yfinance as yf
from sqlalchemy.orm import Session

from config import get_settings
from db.database import SessionLocal
from db.models import StockPrice, Fundamental

logger = logging.getLogger(__name__)
settings = get_settings()


def _upsert_price(session: Session, ticker: str, row: dict) -> None:
    """Insert or update a single price row, avoiding duplicates."""
    existing = (
        session.query(StockPrice)
        .filter(StockPrice.ticker == ticker, StockPrice.date == row["date"])
        .first()
    )
    if existing:
        existing.open = row["open"]
        existing.high = row["high"]
        existing.low = row["low"]
        existing.close = row["close"]
        existing.adj_close = row["adj_close"]
        existing.volume = row["volume"]
    else:
        session.add(StockPrice(
            ticker=ticker,
            date=row["date"],
            open=row["open"],
            high=row["high"],
            low=row["low"],
            close=row["close"],
            adj_close=row["adj_close"],
            volume=row["volume"],
        ))


def fetch_prices(
    ticker: str,
    period: str = "1y",
    session: Optional[Session] = None,
) -> int:
    """
    Fetch historical OHLCV prices for a single ticker from Yahoo Finance.

    Args:
        ticker: Stock symbol (e.g. "AAPL").
        period: yfinance period string ("1d", "5d", "1mo", "3mo", "6mo",
                "1y", "2y", "5y", "10y", "ytd", "max").
        session: SQLAlchemy session. A new one is created if None.

    Returns:
        Number of rows upserted.
    """
    own_session = session is None
    if own_session:
        session = SessionLocal()

    try:
        logger.info("Fetching prices for %s (period=%s)", ticker, period)
        stock = yf.Ticker(ticker)
        df = stock.history(period=period, auto_adjust=False)

        if df.empty:
            logger.warning("No price data returned for %s", ticker)
            return 0

        count = 0
        for idx, row in df.iterrows():
            price_date = idx.date() if hasattr(idx, "date") else idx
            _upsert_price(session, ticker, {
                "date": price_date,
                "open": round(float(row["Open"]), 4) if row["Open"] == row["Open"] else None,
                "high": round(float(row["High"]), 4) if row["High"] == row["High"] else None,
                "low": round(float(row["Low"]), 4) if row["Low"] == row["Low"] else None,
                "close": round(float(row["Close"]), 4) if row["Close"] == row["Close"] else None,
                "adj_close": round(float(row["Adj Close"]), 4) if row["Adj Close"] == row["Adj Close"] else None,
                "volume": int(row["Volume"]) if row["Volume"] == row["Volume"] else None,
            })
            count += 1

        session.commit()
        logger.info("Upserted %d price rows for %s", count, ticker)
        return count

    except Exception as e:
        session.rollback()
        logger.error("Error fetching prices for %s: %s", ticker, e)
        raise
    finally:
        if own_session:
            session.close()


def fetch_fundamentals(
    ticker: str,
    session: Optional[Session] = None,
) -> Optional[Fundamental]:
    """
    Fetch fundamental / valuation data for a single ticker from Yahoo Finance.

    Args:
        ticker: Stock symbol.
        session: SQLAlchemy session. A new one is created if None.

    Returns:
        The upserted Fundamental ORM object, or None on failure.
    """
    own_session = session is None
    if own_session:
        session = SessionLocal()

    try:
        logger.info("Fetching fundamentals for %s", ticker)
        stock = yf.Ticker(ticker)
        info = stock.info

        if not info or info.get("regularMarketPrice") is None:
            logger.warning("No fundamental data returned for %s", ticker)
            return None

        today = date.today()

        # Check for existing record for this ticker + date
        existing = (
            session.query(Fundamental)
            .filter(Fundamental.ticker == ticker, Fundamental.date_fetched == today)
            .first()
        )

        def _safe_float(key: str) -> Optional[float]:
            val = info.get(key)
            if val is None:
                return None
            try:
                return float(val)
            except (TypeError, ValueError):
                return None

        data = {
            "market_cap": _safe_float("marketCap"),
            "pe_ratio": _safe_float("trailingPE"),
            "pb_ratio": _safe_float("priceToBook"),
            "ps_ratio": _safe_float("priceToSalesTrailing12Months"),
            "peg_ratio": _safe_float("pegRatio"),
            "ev_to_ebitda": _safe_float("enterpriseToEbitda"),
            "roe": _safe_float("returnOnEquity"),
            "roa": _safe_float("returnOnAssets"),
            "debt_to_equity": _safe_float("debtToEquity"),
            "current_ratio": _safe_float("currentRatio"),
            "free_cash_flow": _safe_float("freeCashflow"),
            "revenue_growth": _safe_float("revenueGrowth"),
            "earnings_growth": _safe_float("earningsGrowth"),
            "dividend_yield": _safe_float("dividendYield"),
            "sector": info.get("sector"),
            "industry": info.get("industry"),
        }

        if existing:
            for key, value in data.items():
                setattr(existing, key, value)
            fundamental = existing
        else:
            fundamental = Fundamental(
                ticker=ticker,
                date_fetched=today,
                **data,
            )
            session.add(fundamental)

        session.commit()
        logger.info("Upserted fundamentals for %s", ticker)
        return fundamental

    except Exception as e:
        session.rollback()
        logger.error("Error fetching fundamentals for %s: %s", ticker, e)
        raise
    finally:
        if own_session:
            session.close()


def bulk_fetch_prices(
    tickers: list[str],
    period: str = "1y",
    session: Optional[Session] = None,
) -> dict[str, int]:
    """
    Fetch prices for multiple tickers sequentially, sharing a single session.

    Args:
        tickers: List of stock symbols.
        period: yfinance period string.
        session: SQLAlchemy session. A new one is created if None.

    Returns:
        Dict mapping ticker -> number of rows upserted.
    """
    own_session = session is None
    if own_session:
        session = SessionLocal()

    results: dict[str, int] = {}
    try:
        for ticker in tickers:
            try:
                count = fetch_prices(ticker, period=period, session=session)
                results[ticker] = count
            except Exception as e:
                logger.error("Skipping %s due to error: %s", ticker, e)
                results[ticker] = 0
        return results
    finally:
        if own_session:
            session.close()

File: backend/src/ingestion/yahoo.ts
/**
 * Yahoo Finance data ingestion via yahoo-finance2.
 * Fetches historical prices and fundamental data for stocks.
 */

import { PrismaClient } from '@prisma/client';
import { getPrisma } from '../db/client';
import { getSettings } from '../config';
import { startOfDay } from '../utils/format';
import { getYahooFinance } from '../utils/yahooFinance';

const PERIOD_MAP: Record<string, string> = {
  '1mo': '1mo', '3mo': '3mo', '6mo': '6mo',
  '1y': '1y', '2y': '2y', '5y': '5y', '1d': '1d', '5d': '5d',
};

export async function fetchPrices(
  ticker: string,
  period: string = '1y',
  prisma?: PrismaClient,
): Promise<number> {
  const db = prisma || getPrisma();

  console.log(`Fetching prices for ${ticker} (period=${period})`);

  try {
    const now = new Date();
    const periodDays: Record<string, number> = {
      '1d': 1, '5d': 5, '1mo': 30, '3mo': 90, '6mo': 180,
      '1y': 365, '2y': 730, '5y': 1825,
    };
    const days = periodDays[period] || 365;
    const period1 = new Date(now.getTime() - days * 24 * 60 * 60 * 1000);

    const yahooFinance = await getYahooFinance();
    const result = await yahooFinance.historical(ticker, {
      period1,
      period2: now,
    });

    if (!result || result.length === 0) {
      console.warn(`No price data returned for ${ticker}`);
      return 0;
    }

    let count = 0;
    for (const row of result) {
      const priceDate = startOfDay(row.date);

      await db.stockPrice.upsert({
        where: {
          uq_ticker_date: { ticker, date: priceDate },
        },
        update: {
          open: row.open ?? null,
          high: row.high ?? null,
          low: row.low ?? null,
          close: row.close ?? null,
          adjClose: row.adjClose ?? row.close ?? null,
          volume: row.volume ?? null,
        },
        create: {
          ticker,
          date: priceDate,
          open: row.open ?? null,
          high: row.high ?? null,
          low: row.low ?? null,
          close: row.close ?? null,
          adjClose: row.adjClose ?? row.close ?? null,
          volume: row.volume ?? null,
        },
      });
      count++;
    }

    console.log(`Upserted ${count} price rows for ${ticker}`);
    return count;
  } catch (error) {
    console.error(`Error fetching prices for ${ticker}:`, error);
    throw error;
  }
}

export async function fetchFundamentals(
  ticker: string,
  prisma?: PrismaClient,
): Promise<any> {
  const db = prisma || getPrisma();

  console.log(`Fetching fundamentals for ${ticker}`);

  try {
    const yahooFinance = await getYahooFinance();
    const quote = await yahooFinance.quoteSummary(ticker, {
      modules: ['summaryDetail', 'defaultKeyStatistics', 'financialData', 'assetProfile'],
    });

    if (!quote) {
      console.warn(`No fundamental data returned for ${ticker}`);
      return null;
    }

    const sd = quote.summaryDetail || {} as any;
    const ks = quote.defaultKeyStatistics || {} as any;
    const fd = quote.financialData || {} as any;
    const ap = quote.assetProfile || {} as any;

    const todayDate = startOfDay();

    const safeNum = (val: any): number | null => {
      if (val === undefined || val === null) return null;
      const n = Number(val);
      return isNaN(n) ? null : n;
    };

    const data = {
      marketCap: safeNum(sd.marketCap),
      peRatio: safeNum(sd.trailingPE),
      pbRatio: safeNum(ks.priceToBook),
      psRatio: safeNum(sd.priceToSalesTrailing12Months),
      pegRatio: safeNum(ks.pegRatio),
      evToEbitda: safeNum(ks.enterpriseToEbitda),
      roe: safeNum(fd.returnOnEquity),
      roa: safeNum(fd.returnOnAssets),
      debtToEquity: safeNum(fd.debtToEquity),
      currentRatio: safeNum(fd.currentRatio),
      freeCashFlow: safeNum(fd.freeCashflow),
      revenueGrowth: safeNum(fd.revenueGrowth),
      earningsGrowth: safeNum(fd.earningsGrowth),
      dividendYield: safeNum(sd.dividendYield),
      sector: ap.sector || null,
      industry: ap.industry || null,
    };

    // Check for existing record today
    const existing = await db.fundamental.findFirst({
      where: { ticker, dateFetched: todayDate },
    });

    if (existing) {
      await db.fundamental.update({
        where: { id: existing.id },
        data,
      });
    } else {
      await db.fundamental.create({
        data: { ticker, dateFetched: todayDate, ...data },
      });
    }

    console.log(`Upserted fundamentals for ${ticker}`);
    return data;
  } catch (error) {
    console.error(`Error fetching fundamentals for ${ticker}:`, error);
    throw error;
  }
}

export async function bulkFetchPrices(
  tickers: string[],
  period: string = '1y',
  prisma?: PrismaClient,
): Promise<Record<string, number>> {
  const db = prisma || getPrisma();
  const results: Record<string, number> = {};

  for (const ticker of tickers) {
    try {
      results[ticker] = await fetchPrices(ticker, period, db);
    } catch (error) {
      console.error(`Skipping ${ticker} due to error:`, error);
      results[ticker] = 0;
    }
  }

  return results;
}
